relayer/checkpoint.go

```diff
@@ -0,0 +1,212 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package relayer
+
+import (
+	"container/heap"
+	"strconv"
+	"sync"
+
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/awm-relayer/database"
+	"github.com/ava-labs/awm-relayer/utils"
+	"go.uber.org/zap"
+)
+
+//
+// checkpointManager commits keys to be written to the database in a thread safe manner.
+//
+
+type checkpointManager struct {
+	logger                   logging.Logger
+	database                 database.RelayerDatabase
+	writeSignal              chan struct{}
+	relayerID                database.RelayerID
+	queuedHeightsAndMessages map[uint64]*messageCounter
+	committedHeight          uint64
+	lock                     *sync.RWMutex
+	pendingCommits           *utils.UInt64Heap
+	finished                 chan uint64
+}
+
+func newCheckpointManager(logger logging.Logger, database database.RelayerDatabase, writeSignal chan struct{}, relayerID database.RelayerID) *checkpointManager {
+	h := &utils.UInt64Heap{}
+	heap.Init(h)
+	return &checkpointManager{
+		logger:                   logger,
+		database:                 database,
+		writeSignal:              writeSignal,
+		relayerID:                relayerID,
+		queuedHeightsAndMessages: make(map[uint64]*messageCounter),
+		lock:                     &sync.RWMutex{},
+		pendingCommits:           h,
+		finished:                 make(chan uint64),
+	}
+}
+
+func (cm *checkpointManager) run() {
+	go cm.listenForFinishedRelays()
+	go cm.listenForWriteSignal()
+}
+
+func (cm *checkpointManager) writeToDatabase() {
+	cm.lock.RLock()
```

Comment:
Strictly speaking, the read lock isn't necessary here, as long as we copy `cm.committedHeight` and use the same value throughout the function. `cm.committedHeight` is monotonically increasing, and we write periodically and don't care about a little lossiness, so we don't really care about other threads reading/writing to it.

We only have one thread per relayer that can write, but even if we had more, I don't think it would matter.

`database.LatestProcessedBlockKey` is a const so can't be modified. Unless I misinterpreted your question.

Reply:
Actually, to answer most of your questions more succinctly... If two threads have the read lock simultaneously, they necessarily have the same values

Reply:
> For my own understanding: we are acquiring a read lock here to protect fields such as committedHeight, but it is also safe to perform write operations to the database (e.g., cm.database.Put) since the database has its own concurrency control mechanism?

Yes, write concurrency control is assumed to be provided by the `RelayerDatabase` implementation. 

> Two concurrent threads calling writeToDatabase would potentially call cm.database.Put in a different order than their reads to fields in earlier parts of the function. This looks safe, but I just want to confirm.

Within a single `awm-relayer` process, there is a one-to-one mapping between database keys and checkpoint managers, so there should only ever be a single thread writing to a key at any given time. Therefore, the scenario you described should not be possible _within a single `awm-relayer` process_. However, it's possible to deploy a fleet of relayers that all share the same database instance, in which case concurrent writes to the same key are possible. We rely on the database's implementation to provide write consistency in that scenario.

> Can database.LatestProcessedBlockKey be modified by a different concurrent thread? If it can, then it appears we are reading it in our cm.database.Put without providing mutual exclusion unless it is meant to be protected by cm.lock.

Hm that's a good call-out. As mentioned above, only one thread per process is able to access a particular key at a time, but it may be worth adding an `OrderedPut` or a `SafePut` that accepts a predicate and only writes if the predicate passes. We could then more easily atomically perform the check in this funciton. I think there are more considerations here than are worth including in this PR, but I'll create a ticket to track.

Reply:
Thanks for the answers. I think I have a better understanding of the code now. The reason why I asked about performing writes to the DB after acquiring just a read lock is that it is an uncommon pattern. It did look safe, but I just wanted to make sure I wasn't missing anything. 

I didn't realize that database.LatestProcessedBlockKey is a const. I assumed with the word "Latest" in the name that the value would change over time. Perhaps we might want to consider changing its name.

From Cam's answer, it seems that multiple awm-relayer processes can potentially access the same key concurrently (although I assume this is something that we want to avoid in most cases). Adding some type of predicate could be useful as a external check, but I agree it is not something that is critical to add.